File: insertion-sort/typescript/insertion-sort.ts
/**
 * 插入排序
 * @param arr     要排序的数组
 * @param compare 用于排序比较的函数
 */
export default function insertionSort<T>(arr: T[], compare: (a: T, b: T) => number = (a, b) => a > b ? 1 : a < b ? -1 : 0): T[] {
	// 将数值进行复制
	arr = Array.from(arr);
	// n 为数组长度
	const n = arr.length;
	for (let i = 1; i < n; i++) {
		for (let j = i; j > 0; j--) {
			if (compare(arr[j - 1], arr[j]) > 0) {
				// 交换两个数值
				[arr[j], arr[j - 1]] = [arr[j - 1], arr[j]];
			}
		}
	}
	return arr;
}

File: insertion-sort/c/insertion_sort.c
void insertion_sort(int arr[], int n)
{
	for (int i = 1; i < n; i++)
	{
		for (int j = i; j > 0; j--)
		{
			if (arr[j - 1] > arr[j])
			{
				// 交换两个数值
				int temp = arr[j];
				arr[j] = arr[j - 1];
				arr[j - 1] = temp;
			}
			else
			{
				// 不再需要交换，结束此轮循环
				break;
			}
		}
	}
}
